File: 20200705-java-gol/src/main/java/com/katas/Cell.java
package com.katas;

import java.util.List;

public class Cell {
    public static final int MIN_SURVIVAL_THRESHOLD = 2;
    public static final int MAX_SURVIVAL_THRESHOLD = 3;
    public static final int REPRODUCTION_NEIGHBOURS_COUNT = 3;
    private List<Cell> neighbours;
    private boolean isAlive = true;

    public Cell() {
    }

    public Cell(List<Cell> neighbours) {
        this.neighbours = neighbours;
    }

    public Cell(List<Cell> neighbours, boolean isAlive) {
        this.neighbours = neighbours;
        this.isAlive = isAlive;
    }

    public boolean isAlive() {
        return isAlive;
    }

    public void nextGeneration() {
        this.isAlive = isAlive && (neighbours.size() >= MIN_SURVIVAL_THRESHOLD &&
                neighbours.size() <= MAX_SURVIVAL_THRESHOLD)
                || neighbours.size() == REPRODUCTION_NEIGHBOURS_COUNT;
    }
}

File: 20171030-kotlin-game-of-life/src/main/kotlin/com/katas/Cell.kt
package com.katas

class Cell(
        val alive: Boolean = false,
        var neighbours: List<Cell> = mutableListOf<Cell>()
) {
    fun nextGeneration(): Cell {
        if (!alive && aliveNeighboursCount() != 3) return this
        return Cell(aliveNeighboursCount() in 2..3)
    }

    private fun aliveNeighboursCount() = neighbours.fold(0) { acc, cell -> if (cell.alive) acc + 1 else acc }
}
